### DISCUSSION TITLE
Questions about certain things in the scripts

### DISCUSSION THREAD

**Discussion by markusschloesser** [markusschloesser/MackieC4_P3#22] [Q&A]
in [encodercontroller()](https://github.com/markusschloesser/MackieC4_P3/blob/489c46748fbe2fb6417514d22645a0ae64f706cb/EncoderController.py#L729) you have :
```
        encoder_07_index = 6
        encoder_08_index = 7
        encoder_25_index = 24
        encoder_26_index = 25
        encoder_29_index = 28
        encoder_30_index = 29
        encoder_31_index = 30
        encoder_32_index = 31
```

What is it for and why did you do it? I probably studied that for 3hrs total and STILL don't understand it 🙄🤔 @BilldarBagdar

**Comment by BilldarBagdar** [comment]
Edit - removed email garbage, updated some content

To me - it makes sense to be explicit about the difference between the "thing" and the "address" of that thing.  When I think about the physical encoders on the device, I think of them as 1 - 32 in four groups of eight 1 - 8, 9 - 16, 17 - 24, and 25 - 32.  So, "encoders 7 and 8" are the last two on the top row (right side), and that's the "natural way I want to think about those encoders, as 7 and 8."   The "address" (or index) of the encoders (that the computer understands) is of less importance to me.  I didn't want to have to constantly mentally juggle remembering the translation between "thing" and "address of thing" when I re-read that code.  So I created "local variable names" named the way I prefer to think, for the "things", the encoders.  I know what and where "encoder 8" is, its "computer name and location" could be anywhere for all I care. (I only care because the computer won't run correctly if the "computer name and location" are not correctly identified.  But I only need to perform the translation away from my preferred mental map when I specifically need to think about the "computer name and location")

Rather than hard coding the "computer name and location" (the numerical "index values") (which would be considered "magic numbers" in CS "code quality" terms), I gave the raw numbers "better names".  (Another name that also could have replaced the "magic number" 6, for example is "index_6" but IMHO "index_6" isn't really any better than the magic number 6 itself, more verbose with no additional information content.)
  
That's all there is to it really.  You might have a different opinion.  If so, we can talk about it.  But ultimately, all "arguments" of that nature boil down to "who owns the project" and that is you.  I will state my case, as I've done here, if you ask as you've done here.  But if you change one of "my changes" like renaming those kinds of local variables or going back to using the magic numbers, I'm not going to "argue" because ultimately the effort would be futile.  I might ask you why you think your change makes the code more clear, but I'm not "attached to" or "invested in" "my code" in that respect.

I did that kind of thing a lot last year (already?). I also created and named local variables pointing into the "list db" with names that were/are more understandable to me than, for example `self.t_d_p_bank_current[self.t_current]`. (my new names are still not very understandable to me, but are an incremental improvement. IMHO, of course) The "local variables" point to the same location in physical memory as the "list db" location they are assigned to.  The "local variables" are aliases (for the name of the memory location) not copies (of the data stored at the memory location (copy not stored at some other memory location)).  If you change the data at the memory location, "all aliases" still refer to the same location (location didn't change) and (all aliases) will reflect the changed data (when accessed again after the change).

(Also, hopefully you've noticed the pattern in the code where I tried to use "hex notation" like `0x41` when the notation represents a "midi value" and regular decimal notation for all other cases.)
  
Did that answer your questions on this topic?

**Reply by markusschloesser** [reply]
it did indeed!
I'm absolutely fine with it, I just thought there was a deeper meaning for it, which I did not understand.  So that cleared things up. Thank you!
BTW that's the thing that currently is the hardest for me, distinguishing (in the code) between "thing" and "address of thing" as you so aptly put it. That and remembering which bracket/parenthesis does what (list/tuple/set) 🙄😁

>  for example `self.t_d_p_bank_current[self.t_current]`. (my new names are still not very understandable to me, but are an incremental improvement. IMHO, of course)

I struggled A LOT with all the names 😂😂😂 (mostly Leigh's though, I've _kind of_ gotten used to t_d_p now). That's why now a lot has docstrings in the MS_stable commit from yesterday, so I can just hover over things (plisten, omr, oml etc)

**Reply by markusschloesser** [reply]
oh just so I REALLY understand it: `encoder_07_index` represents the stuff that is stored for that specific encoder (in our case param name and param), correct? (cos of "index")

**Reply by BilldarBagdar** [reply]
"Index" has a more general meaning that also applies in this case.  In the method you linked "index" is referring to an "address in a list", but not exactly the address of stuff "stored for the encoder" like device parameters or whatever from Live.  In this case, the "index" is actually just a reference to the physical encoder.  The code indented underneath is both "for the encoder" and "for the associated Live device".

This is how to tell.   Look at how those identifiers are used.

```
""" quoted from master branch at link above"""
            for s in self.__encoders:
                vpot_display_text = ('      ', '      ')
                vpot_param = (None, VPOT_DISPLAY_SINGLE_DOT)
                s_index = s.vpot_index()
                if s_index in row_00_encoders:
                    # if s_index < encoder_07_index:
                    #     always fill the __displayParameters list
                    if s_index == encoder_07_index:
```
The "loop variable" `s` is an item stored in the `__encoders` List.  the `__encoders` List always contains 32 Encoder items, and this list gets "remapped" or "re-initialized" a lot, every time the "mode" changes or selected track or device.  These `s` objects "store stuff" like you mentioned. The parameter and parameter name are examples of properties of Encoder class objects.  Another example would be the "which out of 32 index" this particular `s`, this instance of the Encoder class, is currently "mapped to" across the device. 

Tell me if this is getting too detailed and I will try to come up out of the weeds in the future.  `s_index = s.vpot_index()`  Here, vpot_index() is that "which out of 32 address" **property** of the Encoder class (the instance `s`), and I think I just shortened it to s_index.  But either name still only represents the actual encoder device.  

The next line `if s_index in row_00_encoders:` filters Encoder instances (every time through the loop `s` represents a different encoder) to only pass into the "indented code block" if the encoder is in the first row (aside: I just realized I didn't consistently follow my own naming convention, the const range "row_00_encoders" represents the first row of four (the row at the top) Maybe all those const "encoder row range names" should increment and start at "row_01_encoders" [oopsie])  ...Then the last line quoted here `if s_index == encoder_07_index:` filters for exactly one `s` instance of Encoder, the instance when `s_index == encoder_07_index`.  And then the next indented line (not quoted above) `if current_device_bank_track > 0:` only applies when that one `s` object representing that one encoder (encoder 7) is the current "list item being iterated" (technically I guess the list is being iterated over and the `s` objects are the "objects of iteration").

> somethings wrong here, for tracks which do not show devices (but should) "<<Bank" is displayed

Based on what I just explained, how would any of the filtering that leads from the top of the loop into the "indentation block" representing this one encoder "understand" anything about tracks and devices necessarily?  I don't mean to seem sarcastic or anything rude.  It's more like "Socratic method" time for questions.  We're inside a method called "reassign encoder parameters" and this text is hard coded in the display variable `vpot_display_text = ('<<Bank', 'Device') ` when `current_device_bank_track > 0`.

This is a tricky thing about programming.  When you are reading code, you never know what value will actually be represented by, for example, `current_device_bank_track`.  When you are reading code, what is written could seem to make perfectly logical sense.  But the proof is in the pudding as they say, and the real arbiter of truth is your running program with a debugger attached or emitting log messages.  On the surface one would think `current_device_bank_track > 0` is never true "for tracks which do not show devices (but should)".  That it would only be true when a track has a device and the "current selected parameter" of that device is "beyond" bank 1 (bank 1 is the first 24 automation parameters of a device. bank 1 has the "address" 0)   If something is not true when you think it should be, or vice versa, then it's time to put on your "debugging helmet" and work your way into the darkness.  The idea that code executes (or tries to execute) is that the C4 should only show the "Page To Previous Bank" << display when it is possible (i.e. from bank 2 to bank 1).  Similarly for encoder 8, it is only supposed to show you can "Page to Next Bank" if it is possible.  However, this is only the "reassign parameters" method.  Later over in "on update display timer", the same pattern is seen, looping/iterating over a list.  But this time it is not the list of Encoder class object.  This time the "iteration variable" `t` is from the const "encoder_range" range. (and for this quote we are in "channel strip" mode)

```
            for t in encoder_range:

                text_for_display = self.__display_parameters[t]
                u_alt_text = ' '
                l_alt_text = ' '
                if len(text_for_display) > 1:
                    u_alt_text = text_for_display[1]
                    l_alt_text = text_for_display[0]
                elif len(text_for_display) > 0:
                    l_alt_text = text_for_display[0]

                if t in range(6, NUM_ENCODERS_ONE_ROW):
                    upper_string1 += self.__transform_to_size(u_alt_text, 6)
                    upper_string1 += ' '
                    lower_string1 += self.__transform_to_size(unicode(l_alt_text), 6)
                    lower_string1 += ' '
                elif t in row_01_encoders:
                    lower_string2 += self.__transform_to_size(unicode(l_alt_text), 6)
                    lower_string2 += ' '
                elif t in row_02_encoders:
```
`text_for_display = self.__display_parameters[t]` and the next few lines is code I added to guard against log vomit. IIRC, originally the code was directly "transforming" whatever was in (or not) `self.__display_parameters[t]`

`if t in range(6, NUM_ENCODERS_ONE_ROW):`  means specifically, encoders 7 and 8 and the code indented underneath the if filtering condition means the LCD over the encoders should show whatever is stored in the `self.__display_parameters` list at index `t` as the "upper and lower" text.

Then later, when "plugin mode" is selected `if self.__chosen_plugin is None:` means if **the** "chosen plugin" doesn't exist (is null), the entire LCD row should display "No devices inserted on this track"  (I kind of recall adding this code too, just for "logical completeness" meaning I don't think it should ever be seen on the LCD, but the code still exists because logic is funny that way.  If I never see what I coded and don't expect to see, good.  But if it happens and I didn't code for it, it could easily take "forever" to figure out the problem.)  (See it?  Why would the device be in "plugin mode" when `self.__chosen_plugin is None:` It shouldn't... But can you guarantee it will never happen?  and if it ever happens, would it be better to get an ALL CAPS notification that it happened right on the display, or to get some other result like misleading log vomit?) (Also note that, I added the code while debugging.  The code you see reflects "more than one" debugging session last year.  I didn't just "decide" that code was needed, the log kept filling with vomit and I kept changing stuff until the log spewing stopped.) 

The rest of the code when "plugin mode" is active looks a lot like the for loop quoted above, except only the "lower string1" variable is assigned a value to display for encoders 7 and 8.
